### PR TITLE
Project Hummingbird: Move Account Creation Logic into Domains

### PR DESCRIPTION
**Pull request by kenlewis-sfdo**
WI: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07AH000000Q6jMYAS/view

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: src/classes/Contacts.cls
Comment (lparrott) [inline src/classes/Contacts.cls:207]: Maybe it's just me, but having descriptions inside methods feels weird 🤷‍♂️ could make them block comments, or move them above the method name?

File: src/classes/Contacts.cls
Comment (lparrott) [inline src/classes/Contacts.cls:740]: Out of scope if we're just trying to reorganize, but this could just be a loop over a list of field names.

File: src/classes/HouseholdAccounts.cls
Comment (lparrott) [inline src/classes/HouseholdAccounts.cls:226]: Any chance we already have these Contacts queried? Feels risky to run a query (2 actually...) each time this method is called.

File: src/classes/HouseholdAccounts.cls
Comment (lparrott) [inline src/classes/HouseholdAccounts.cls:241]: I feel like these records should already be available?

File: src/classes/HouseholdAccounts.cls
Comment (lparrott) [inline src/classes/HouseholdAccounts.cls:296]: More queries, maybe this is a larger problem.

File: src/classes/Households.cls
Comment (lparrott) [inline src/classes/Households.cls:502]: Super small thing, could this be an `else if`, since you can only have one Account model active?

**Review by lparrott — COMMENTED** [review]
Looking great, I just have a few comments after a quick run through.

File: src/classes/HouseholdAccounts.cls
Comment (bdvorachek) [reply src/classes/HouseholdAccounts.cls:241]: I think we can combine these queries, my only reservation was "What happens when one of the Accounts has no Contacts?". When looping through the second query, contactIds would be null for any Account that had zero contacts, and that looks to be OK. 

We'll still need a second loop here, but we wouldn't need to query Accounts. We can gather all of the AccountIds in the first loop w/ their `npe01__One2OneContact__c`, and add those to the Accounts to update if the Contact Id in `npe01__One2OneContact__c` does not match with any of the Contacts that are currently part of the given Household.

File: src/classes/Households.cls
Comment (bdvorachek) [inline src/classes/Households.cls:202]:
```suggestion
    public static List<Contact> contactsWithoutAccounts(Contacts contactsInstance) {
```

`oldContactById` is equivalent to `contactInstance.oldMap` here I think, so we could drop the second argument.

File: src/classes/Households.cls
Comment (bdvorachek) [inline src/classes/Households.cls:210]:
```suggestion
                Contact oldContact = contactsInstance.oldMap.get(contactRecord.Id);
```

File: src/classes/Contacts.cls
Comment (kenlewis-sfdo) [reply src/classes/Contacts.cls:207]: Agreed.  Do you think this is a comment that is useful enough to keep? Or should it just be removed?  Not sure that it is worth keeping.

File: src/classes/HouseholdAccounts.cls
Comment (kenlewis-sfdo) [reply src/classes/HouseholdAccounts.cls:296]: This method was fully lifted from ACCT_IndividualAccounts_TDTM, and HouseholdAccounts is evolving from its original incarnation as a way to lazy load a Households' members.  

Now that we have these methods in the same domain, we should absolutely be able to take advantage and query what is needed up front at instantiation, so this type of multiple query pattern can be avoided.

I'll take a look today and see if this seems to make sense to do now.

File: src/classes/Contacts.cls
Comment (sokosaurus) [inline src/classes/Contacts.cls:649]: At some point we may want to consider moving this into a class-level constant variable (or somewhere else) so that we can easily track/maintain a potentially growing list of deprecated fields that need to be nulled out here.

File: src/classes/Contacts.cls
Comment (sokosaurus) [inline src/classes/Contacts.cls:656]: Just a small cleanup item (for the future) - we could consider dropping `== true` from this conditional.

File: src/classes/Contacts.cls
Comment (sokosaurus) [inline src/classes/Contacts.cls:278]: In our variable and method names that are related to the 1:1 model, we use the naming convention of either `OneToOne` or `One2One`. For searchability (and perhaps other benefits), do you think it makes sense to align on a singular naming convention here?

File: src/classes/HouseholdAccounts.cls
Comment (sokosaurus) [inline src/classes/HouseholdAccounts.cls:279]: Probably out of scope for this initial effort, but flagging this as a long running method that we should consider refactoring.

**Review by sokosaurus — CHANGES_REQUESTED** [review]
Took a pass at this PR. Feel free to push back on anything that doesn't feel actionable in this iteration of the work. Looking really good @kenlewis-sfdo!

File: src/classes/Contacts.cls
Comment (kenlewis-sfdo) [reply src/classes/Contacts.cls:578]: Yes, good call.  Let's do that when we have another field to add here.

File: src/classes/Contacts.cls
Comment (kenlewis-sfdo) [reply src/classes/Contacts.cls:207]: Yep.  Should prob standardize on OneToOne, oneToOne, One-to-one, etc.

File: src/classes/HouseholdAccounts.cls
Comment (kenlewis-sfdo) [reply src/classes/HouseholdAccounts.cls:279]: 👍🏽

File: src/classes/HouseholdAccounts.cls
Comment (kenlewis-sfdo) [reply src/classes/HouseholdAccounts.cls:296]: I really like these ideas about combining queries in this class and getting it instantiated with the right set of records.  But going to hold off on that for a follow up next WI: https://salesforce.quip.com/n2vzAQOiRbT7#UMTACAR9Raq

File: src/classes/HouseholdAccounts.cls
Comment (kenlewis-sfdo) [reply src/classes/HouseholdAccounts.cls:279]: Added to design doc for further discussion!: https://salesforce.quip.com/n2vzAQOiRbT7#UMTACAjM6Uc

File: src/classes/HouseholdAccounts.cls
Comment (kenlewis-sfdo) [reply src/classes/HouseholdAccounts.cls:241]: I like the plan, Ben, adding it to the list of items to follow up on on this class: https://salesforce.quip.com/n2vzAQOiRbT7#UMTACAd3ibZ

File: src/classes/Contacts.cls
Comment (lparrott) [reply src/classes/Contacts.cls]: I agree, just delete them (I see 3 areas that are like this)

**Review by lparrott — APPROVED** [review]
Looking good, approving with the assumption that we'll tackle the feedback in a future PR 👍

**Review by sokosaurus — APPROVED** [review]
Following @lparrott and approving. Really great work here!

**Pull request merged**
